### PR TITLE
Documentation - tensorflow_backend

### PR DESCRIPTION
**Pull request by keunwoochoi**
* about the first 1/3. 
* Style: following common.py
* All examples are tested with tf backend.
* Used 'Keras tensor' and 'Keras variable' and tried to make their differences clear, mainly relying on trial-and-error, so there could be errors. 
* 77ae504 contains all changes.

### CODE REVIEW

**Comment by keunwoochoi** [comment]
one pep8 error - will fix with following PR.

**Comment by jmhessel** [comment]
Hi @keunwoochoi -- just to double check, this is the PR that fixes #4588?

**Comment by keunwoochoi** [comment]
Yes, briefly mentioned in this line
https://github.com/fchollet/keras/pull/4677/commits/56c33a98e9ff30775a3fc9e2be29647b5387add9#diff-a18b8c6a1191d6f49303e0d599ca8c37R785 . It might be too brief though, it took me more than a second to find it.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -257,20 +385,78 @@ def ndim(x):
...
+        dtype: String, data type of returned Keras variable
+        name: String, name of returned Keras variable
+
+    # Returns
+        A Keras variable, filled with `0.0`
+
+    # Example
+    ```python
+        >>> from keras import backend as K
+        >>> var = K.zeros((3,4))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+        dtype: String, data type of returned Keras variable.
+        name: String, name of returned Keras variable.
+
+    # Returns
+        A Keras variable, an identity matrix.
+
+    # Example
+    ```python
+        >>> from keras import backend as K
+        >>> K.eval(var)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -279,7 +465,25 @@ def zeros(shape, dtype=_FLOATX, name=None):
...
+        dtype: String, data type of returned Keras variable.
+        name: String, name of returned Keras variable.
+
+    # Returns
+        A Keras variable, filled with `1.0`.
+
+    # Example
+    ```python
+        >>> from keras import backend as K
+        >>> var = K.ones((3,4))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -257,20 +385,78 @@ def ndim(x):
...
+    # Arguments
+        x: A Keras variable.
+
+    # Returns
+        A Numpy array.
+
+    # Examples
+    ```python
+        >>> from keras import backend as K
+        >>> var = K.variable(np.array([[1, 2], [3, 4]]), dtype='float32')
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -257,20 +385,78 @@ def ndim(x):
...
+        >>> from keras import backend as K
+        # Keras tensor
+        >>> K.dtype(K.placeholder(shape=(2,4,5)))
+        'float32'
+        >>> K.dtype(K.placeholder(shape=(2,4,5), dtype='float32'))
+        'float32'
+        >>> K.dtype(K.placeholder(shape=(2,4,5), dtype='float64'))
+        'float64'
+        # Keras variable
+        >>> var = K.variable(np.array([[1, 2], [3, 4]]))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -230,22 +297,83 @@ def placeholder(shape=None, ndim=None, dtype=_FLOATX, sparse=False, name=None):
...
+
+    # Returns
+        integer (scalar), number of dimensions.
+
+    # Examples
+    ```python
+        >>> from keras import backend as K
+        >>> input = K.placeholder(shape=(2, 4, 5))
+        >>> val = np.array([[1, 2], [3, 4]])
+        >>> var = K.variable(value=val)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 7]
```diff
@@ -230,22 +297,83 @@ def placeholder(shape=None, ndim=None, dtype=_FLOATX, sparse=False, name=None):
...
+        A tuple of integers.
+
+    # Examples
+    ```python
+        >>> from keras import backend as K
+        >>> input = K.placeholder(shape=(2, 4, 5))
+        >>> K.int_shape(input)
+        (2, 4, 5)
+        >>> val = np.array([[1, 2], [3, 4]])
+        >>> var = K.variable(value=val)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 8]
```diff
@@ -230,22 +297,83 @@ def placeholder(shape=None, ndim=None, dtype=_FLOATX, sparse=False, name=None):
...
+    # Returns
+        A symbolic shape.
+
+    # Examples
+    ```
+        # TensorFlow example
+        >>> from keras import backend as K
+        >>> tf_session = K.get_session()
+        >>> val = np.array([[1, 2], [3, 4]])
+        >>> var = K.variable(value=val)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 9]
```diff
@@ -150,33 +154,86 @@ def _to_tensor(x, dtype):
...
 
     # Returns
-        Tensor variable instance.
+        A Keras variable instance.
+
+    # Examples
+    ```python
+        >>> from keras import backend as K
+        >>> val = np.array([[1, 2], [3, 4]])
+        >>> var = K.variable(value=val, dtype='float64', name='example_var')
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: var is a reserved keyword, use `v` or something like that.

File: keras/backend/tensorflow_backend.py [hunk 10]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+    # Arguments
+        x: Keras variable.
+
+    # Returns
+        A Keras variable, filled with `0.0`.
+
+    # Example
+    ```python
+        >>> from keras import backend as K
+        >>> var = K.variable(np.random.random((2,3)))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 11]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+    # Arguments
+        x: Keras variable.
+
+    # Returns
+        A Keras variable, filled with `1.0`.
+
+    # Example
+    ```python
+        >>> from keras import backend as K
+        >>> var = K.variable(np.random.random((2,3)))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 12]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+        name: String, name of returned Keras variable.
+        seed: Integer, random seed.
+
+    # Returns
+        A Keras variable, filled with drawn samples.
+
+    # Example
+    ```python
+        # TensorFlow example
+        >>> var = K.random_uniform_variable((2,3), 0, 1)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 13]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+               [ 1.,  1.,  1.]], dtype=float32)
+    ```
     '''
     return tf.ones_like(x, name=name)
 
 
 def random_uniform_variable(shape, low, high, dtype=_FLOATX,
                             name=None, seed=None):
+    '''Instantiates an Keras variable filled with
+    drawn samples from a uniform distribution and returns it.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: "samples drawn from"

File: keras/backend/tensorflow_backend.py [hunk 14]
```diff
@@ -321,6 +600,31 @@ def random_uniform_variable(shape, low, high, dtype=_FLOATX,
 
 def random_normal_variable(shape, mean, scale, dtype=_FLOATX,
                            name=None, seed=None):
+    '''Instantiates an Keras variable filled with
+    drawn samples from a normal distribution and returns it.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: "samples drawn from"

File: keras/backend/tensorflow_backend.py [hunk 15]
```diff
@@ -321,6 +600,31 @@ def random_uniform_variable(shape, low, high, dtype=_FLOATX,
...
+        name: String, name of returned Keras variable.
+        seed: Integer, random seed.
+
+    # Returns
+        A Keras variable, filled with drawn samples.
+
+    # Example
+    ```python
+        # TensorFlow example
+        >>> var = K.random_normal_variable((2,3), 0, 1)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 16]
```diff
@@ -332,14 +636,55 @@ def random_normal_variable(shape, mean, scale, dtype=_FLOATX,
...
+
+    # Arguments
+        x: Keras variable.
+
+    # Returns
+        Integer, the number of scalars in `x`.
+
+    # Example
+    ```python
+        >>> var = K.zeros((2,3))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 17]
```diff
@@ -332,14 +636,55 @@ def random_normal_variable(shape, mean, scale, dtype=_FLOATX,
...
+               [ 0.,  0.,  0.]], dtype=float32)
+    ```
     '''
     shape = x.get_shape()
     return np.prod([shape[i]._value for i in range(len(shape))])
 
 
 def cast(x, dtype):
-    '''Casts a tensor to a different dtype.
+    '''Casts a Keras variable to a different dtype.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: This is primarily meant for tensor casting, not variable casting. Albeit you could cast a variable (thus returning a tensor).

File: keras/backend/tensorflow_backend.py [hunk 18]
```diff
@@ -332,14 +636,55 @@ def random_normal_variable(shape, mean, scale, dtype=_FLOATX,
...
+    # Arguments
+        x: Keras variable.
+        dtype: String, either (`'float16'`, `'float32'`, or `'float64'`).
+
+    # Returns
+        Keras variable with dtype of `dtype`.
+
+    # Example
+    ```python
+        >>> var = K.zeros((2,3), dtype='float32')
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: same

File: keras/backend/tensorflow_backend.py [hunk 19]
```diff
@@ -367,10 +712,47 @@ def moving_average_update(variable, value, momentum):
...
 
 def dot(x, y):
-    '''Multiplies 2 tensors.
+    '''Multiplies 2 tensors and/or variables.
     When attempting to multiply a ND tensor
     with a ND tensor, reproduces the Theano behavior
     (e.g. (2, 3).(4, 3, 5) = (2, 4, 5))
+
+    # Arguments
+        x: Keras variable.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: tensors!

File: keras/backend/tensorflow_backend.py [hunk 20]
```diff
@@ -332,14 +636,55 @@ def random_normal_variable(shape, mean, scale, dtype=_FLOATX,
...
     shape = x.get_shape()
     return np.prod([shape[i]._value for i in range(len(shape))])
 
 
 def cast(x, dtype):
-    '''Casts a tensor to a different dtype.
+    '''Casts a Keras variable to a different dtype.
+
+    # Arguments
+        x: Keras variable.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Tensor!

File: keras/backend/tensorflow_backend.py [hunk 21]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
 
 
 def ones_like(x, name=None):
-    '''Instantiates an all-ones tensor
-    of the same shape as another tensor.
+    '''Instantiates an all-ones Keras variable
+    of the same shape as another Keras variable and returns it.
+
+    # Arguments
+        x: Keras variable.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: or tensor

File: keras/backend/tensorflow_backend.py [hunk 22]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+    ```
     '''
     return tf.zeros_like(x, name=name)
 
 
 def ones_like(x, name=None):
-    '''Instantiates an all-ones tensor
-    of the same shape as another tensor.
+    '''Instantiates an all-ones Keras variable
+    of the same shape as another Keras variable and returns it.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: or tensor

File: keras/backend/tensorflow_backend.py [hunk 23]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
 
 
 def zeros_like(x, name=None):
-    '''Instantiates an all-zeros tensor
-    of the same shape as another tensor.
+    '''Instantiates an all-zeros Keras variable
+    of the same shape as another Keras variable and returns it.
+
+    # Arguments
+        x: Keras variable.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: or tensor

File: keras/backend/tensorflow_backend.py [hunk 24]
```diff
@@ -288,27 +492,102 @@ def ones(shape, dtype=_FLOATX, name=None):
...
+
     '''
     return variable(np.eye(size), dtype, name)
 
 
 def zeros_like(x, name=None):
-    '''Instantiates an all-zeros tensor
-    of the same shape as another tensor.
+    '''Instantiates an all-zeros Keras variable
+    of the same shape as another Keras variable and returns it.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: or tensor

File: keras/backend/tensorflow_backend.py [hunk 25]
```diff
@@ -367,10 +712,47 @@ def moving_average_update(variable, value, momentum):
...
+        >>> xy = K.dot(x, y)
+        >>> K.int_shape(xy)
+        (2, 4)
+        >>> K.eval(xy)
+        array([[ 0.36257121,  0.450986  ,  0.49147484,  0.20764332],
+               [ 0.45370784,  1.53120756,  1.37138271,  0.24922732]], dtype=float32)
+    ```
+
+    ```python
+        # dot product between Keras tensor and Keras variable
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Not a useful example. Instead please have examples dealing with larger input shapes (e.g. 3D, 4D).

**Comment by fchollet** [comment]
Also, there seems to be merge conflicts now.

File: keras/backend/tensorflow_backend.py (hunk 25 above)
Comment (keunwoochoi) [inline keras/backend/tensorflow_backend.py]: Can't think of the right example, I'll just change `x = K.random_uniform_variable(shape=(2, 3), low=0, high=1)` to `x = K.placeholder(shape=(2, 3))` (same for y).

**Comment by keunwoochoi** [comment]
Great, will work on the rest of backend soon.

File: keras/backend/tensorflow_backend.py (hunk 9 above)
Comment (keunwoochoi) [inline keras/backend/tensorflow_backend.py]: done.

**Comment by keunwoochoi** [comment]
All requested changes are done with 5fa311c.

**Comment by fchollet** [comment]
Good, thanks!

**Pull request merged**